synthtool/languages/python_mono_repo.py

```diff
@@ -98,6 +99,54 @@ def update_url_in_setup_py(package_dir: str):
         f.writelines(new_setup_py)
 
 
+def apply_client_specific_post_processing(client_specific_post_processing_dir: Path):
+    """Applies client specific post processing which exists in
+    the Path `client_specific_post_processing_dir`. The client specific post processing
+    YAML is in the following format:
+    ```
+        description: Verbose description about the need for the workaround.
+        url: Url of the issue in gapic-generator-python.
+        replacements:
+          - replacement:
+            paths: [<List of files where the replacement should occur>]
+            before: ""The string to search for in the specified paths""
+            after:  "The string to replace in the the specified paths",
+            count: <integer indicating number of replacements that should have occurred across all files after the script is run>
+    ```
+
+    Args:
+        client_specific_post_processing_dir (str): path to the directory which contains YAML files which will
+            be used to apply client specific post processing. For example, 'scripts/client-post-processing'
+    """
+    if Path(client_specific_post_processing_dir).exists():
+        for post_processing_path in Path(client_specific_post_processing_dir).iterdir():
+            with open(post_processing_path, "r") as post_processing_path_file:
+                post_processing_json = yaml.safe_load(post_processing_path_file)
+                replacements = post_processing_json["replacements"]
+
+                # For each workaround related to the specified issue
+                for replacement in replacements:
+                    # For each file that needs the workaround applied
+                    for client_library_path in replacement["paths"]:
+                        assert (
+                            s.replace(
+                                client_library_path,
+                                replacement["before"],
+                                replacement["after"],
+                            )
+                            == replacement["count"]
```

Comment:
Doesn't this violate idempotency? We'll make the replacements in all the package dirs the first time. Then when this script is triggered again, we'll only make changes in the dirs that changed; the other ones will get no changes, so there will be no replacements, so the assertion will fail.

In the design doc we said that `replacement["count"]` will be used to check the difference between the original and changed files, so it seems like we need to somehow cache the original file before it's ever processed by this script, and then we can compare it to the what it becomes after this script has run any number of times.

Maybe you mean to only selectively trigger this script? But then perhaps it would make sense to implement the symlink idea I mentioned in the design doc. (My thinking: if the workaround file applies to packages A and B and only B has new changes, the assertion will fail on A.)

Happy to discuss more, especially if my interpretation/mental model is incorrect.

Reply:
You're right. I meant to trigger the replacement selectively. I've updated the design doc to include symlinks under package specific folders, and implemented the idea in this PR. Please take another look at the design doc and changes in [c525b70](https://github.com/googleapis/synthtool/pull/1881/commits/c525b700e68b22e19fda4ae5544638d9dab2fe82).

Reply:
Resolving old thread